src/Chainweb/Pact/Backend/Utils.hs

```diff
@@ -0,0 +1,204 @@
+  deriving (Eq, Ord, Enum)
+
+instance Show SavepointName where
+  show (Block) = "block"
+  show (DbTransaction) = "db-transaction"
+  show (PreBlock) = "preblock"
+  show (PactDbTransaction) = "pact-transaction"
+
+instance AsString SavepointName where
+  asString = Data.Text.pack . show
+
+withBlockEnv :: a -> (MVar a -> IO b) -> IO b
+withBlockEnv blockenv f = newMVar blockenv >>= f
+
+readBlockEnv :: (a -> b) -> MVar a -> IO b
+readBlockEnv f = fmap f . readMVar
+
+withSQLiteConnection :: String -> [Pragma] -> Bool -> (SQLiteEnv -> IO c) -> IO c
+withSQLiteConnection file ps todelete action = bracket opener closer action
+  where
+    closer c = do
+      void $ close_v2 $ _sConn c
+      when todelete (removeFile file)
+    opener = do
+      e <- open_v2 (fromString file) (0x00000002 .|. 0x00000004 .|. 0x00010000) "unix"
+      case e of
+        Left (err, msg) ->
+          internalError $
+            "withSQLiteConnection: Can't open db with "
+            <> asString (show err) <> ": " <> asString (show msg)
+        Right r -> do
+          runPragmas r ps
+          return $ mkSQLiteEnv r
+    mkSQLiteEnv connection =
+      SQLiteEnv connection
+      (SQLiteConfig file ps)
+
+withTempSQLiteConnection :: [Pragma] -> (SQLiteEnv -> IO c) -> IO c
+withTempSQLiteConnection ps action =
+  withTempFile (\file -> withSQLiteConnection file ps False action)
+
+domainTableName :: Domain k v -> Utf8
+domainTableName = Utf8 . toS . asString
+
+convKeySetName :: KeySetName -> Utf8
+convKeySetName (KeySetName name) = Utf8 $ toS name
+
+convModuleName :: ModuleName -> Utf8
+convModuleName (ModuleName name _) = Utf8 $ toS name
+
+convNamespaceName :: NamespaceName -> Utf8
+convNamespaceName (NamespaceName name) = Utf8 $ toS name
+
+convRowKey :: RowKey -> Utf8
+convRowKey (RowKey name) = Utf8 $ toS name
+
+convPactId :: PactId -> Utf8
+convPactId = Utf8 . toS . show
+
+expectSingleRowCol :: Show a => String -> [[a]] -> IO a
+expectSingleRowCol _ [[s]] = return s
+expectSingleRowCol s v =
+  internalError $
+  "expectSingleRowCol: "
+  <> asString s <>
+  " expected single row and column result, got: "
+  <> asString (show v)
+
+expectSingle :: Show a => String -> [a] -> IO a
+expectSingle _ [s] = return s
+expectSingle desc v =
+  internalError $
+  "Expected single-" <> asString (show desc) <> " result, got: " <>
+  asString (show v)
+
+
+instance StringConv Text Utf8 where
+  strConv l = Utf8 . strConv l
+
+instance StringConv Utf8 Text where
+  strConv l (Utf8 bytestring) = strConv l bytestring
+
+instance StringConv ByteString Utf8 where
+  strConv l = Utf8 . strConv l
+
+instance StringConv Utf8 ByteString where
+  strConv l (Utf8 bytestring) = strConv l bytestring
+
+instance StringConv String Utf8 where
+  strConv l = Utf8 . strConv l
+
+instance StringConv Utf8 String where
+  strConv l (Utf8 bytestring) = strConv l bytestring
+
+fastNoJournalPragmas :: [Pragma]
+fastNoJournalPragmas = [
+  "synchronous = NORMAL",
+  "journal_mode = WAL",
```

Comment:
Why WAL and not MEMORY?

Reply:
`MEMORY` isn't safe (if machine dies, process crashes or is unsafely stopped, etc., the DB will be corrupt.) We do need correct transactional semantics here